hw/mcu/nordic/nrf52xxx/src/hal_watchdog.c

```diff
@@ -44,13 +44,20 @@ nrf52_wdt_irq_handler(void)
 int
 hal_watchdog_init(uint32_t expire_msecs)
 {
-    uint64_t expiration;
-
     NRF_WDT->CONFIG = WDT_CONFIG_SLEEP_Msk;
 
-    /* Convert msec timeout to counts of a 32768 crystal */
-    expiration = ((uint64_t)expire_msecs * 32768) / 1000;
-    NRF_WDT->CRV = (uint32_t)expiration;
+    if (expire_msecs >= 1048576) {
+        /* watchdog with more than 17 minutes? */
+        assert(0);
+    } else {
+        uint32_t e = expire_msecs, rtc = 32768, d = 1000;
+        while (e >= 131072) {
+            e /= 2;
+            rtc /= 2;
+            d /= 2;
```

Comment:
I think you could do this in the following manner and get much more range. I probably should have done this from the  start but did not worry about the 65-bit divide. We are trying to do the following: (x * 32768) / 1000. This is the same as doing this: x * 32.768. Since we do not want to use floating point you can do this: (x * 32) + ((x * 768) / 1000) which is also this: (x * 32) + ((x * 96) / 125). You can then get much more range.

Reply:
Cool!

```
#include <stdint.h>
#include <stdio.h>

int main(void)
{
    uint64_t r;
    uint32_t s;

    for (uint32_t u = 0; u < 0xffffffff; u++) {
        r = ((uint64_t)u * 32768) / 1000;
        s = (u * 32) + ((u * 96) / 125);

        if (r != s) {
            printf("u=%u, r=%lu, s=%u\n", u, r, s);
            break;
        }
    }

    return 0;
}
```

```
$ ./test
u=44739243, r=1466015514, s=1431655776
```

I guess you just made it work upto 12 hours, you're a genius @wes3!

Reply:
Thanks Fabio but not really warranted.